File: 85-maximal-rectangle/maximal-rectangle.java
class Solution {

    private int largestHistogram(int[] heights) {
        Stack<Integer> stack = new Stack();
        int maxArea = 0;
        int n = heights.length;
        
        for (int i = 0; i <= n; i++) {
            int height = (i == n) ? 0 : heights[i];
            while (!stack.isEmpty() && heights[stack.peek()] > height) {
                int h = heights[stack.pop()];
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                maxArea = Math.max(maxArea, h * width);
            }
            stack.push(i);
        }
        return maxArea;
    }

    public int maximalRectangle(char[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[] heights = new int[cols];
        int maxArea = 0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                heights[c] = (matrix[r][c] == '1') ? heights[c] + 1 : 0;
            }
            maxArea = Math.max(maxArea, largestHistogram(heights));
        }
        return maxArea;
    }
}

File: 85-maximal-rectangle/maximal-rectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        # find largest histogram at every row
        rows, cols = len(matrix), len(matrix[0])

        def largest_histogram(heights):
            stack, largest = [], 0
            for i, height in enumerate(heights + [0]):
                start = i
                while stack and stack[-1][0] > height:
                    prev_h, start = stack.pop()
                    largest = max(largest, prev_h * (i - start))
                stack.append((height, start))
            return largest

        maximal_rect = 0
        prev = [0] * cols
        for r in range(rows):
            nxt = [0] * cols
            for c in range(cols):
                nxt[c] = prev[c] + 1 if matrix[r][c] == '1' else 0
            maximal_rect = max(maximal_rect, largest_histogram(nxt))
            # print(nxt, largest_histogram(nxt))
            prev = nxt

        return maximal_rect